arangod/RestHandler/RestMetricsHandler.cpp

```diff
@@ -79,85 +85,184 @@ RestStatus RestMetricsHandler::execute() {
+    //             I think we should fix it, but its breaking change
+    generateError(
+        notFound ? rest::ResponseCode::NOT_FOUND : rest::ResponseCode::BAD,
+        TRI_ERROR_HTTP_BAD_PARAMETER, error);
+    return RestStatus::DONE;
+  }
+
+  if (foundServerId) {
+    return makeRedirection(serverId, false);
+  }
+
+  if (type == "cd_json") {
+    auto& metrics = server().getFeature<metrics::ClusterMetricsFeature>();
+    auto data = metrics.getData();
+    _response->setResponseCode(rest::ResponseCode::OK);
+    _response->setContentType(rest::ContentType::VPACK);
+    if (data->packed) {
+      _response->addPayload(velocypack::Slice{data->packed->data()});
+    } else {
+      _response->addPayload(velocypack::Slice::emptyArraySlice());
+    }
+    return RestStatus::DONE;
+  }
+
   auto& metrics = server().getFeature<metrics::MetricsFeature>();
   if (!metrics.exportAPI()) {
     // don't export metrics, if so desired
     generateError(rest::ResponseCode::NOT_FOUND, TRI_ERROR_HTTP_NOT_FOUND);
     return RestStatus::DONE;
   }
 
-  auto const& values = _request->values();
-  auto it = values.find("type");
-  if (it != values.end() && it->second == "json") {
+  if (type == "db_json") {
     VPackBuilder builder;
     metrics.toVPack(builder);
     _response->setResponseCode(rest::ResponseCode::OK);
     _response->setContentType(rest::ContentType::VPACK);
     _response->addPayload(builder.slice());
-  } else if (it != values.end()) {
-    generateError(rest::ResponseCode::NOT_FOUND, TRI_ERROR_HTTP_BAD_PARAMETER,
-                  std::string("unknown 'type' parameter supplied."));
     return RestStatus::DONE;
-  } else {
+  }
+
+  auto const leader = [&]() -> std::optional<std::string> {
+    auto& cm = server().getFeature<metrics::ClusterMetricsFeature>();
+    if (cm.isEnabled() && mode != metrics::CollectMode::Local) {
+      return cm.update(mode);
+    }
+    return std::nullopt;
+  }();
+
+  if (leader && (leader->empty() || type == "last")) {
+    // TODO(MBkkt) Maybe response with some error?
+    mode = metrics::CollectMode::Local;
+  }
+
+  if (!leader) {
     std::string result;
-    metrics.toPrometheus(result);
+    metrics.toPrometheus(result, mode);
     _response->setResponseCode(rest::ResponseCode::OK);
     _response->setContentType(rest::ContentType::TEXT);
     _response->addRawPayload(result);
+    return RestStatus::DONE;
+  }
+  TRI_ASSERT(mode == metrics::CollectMode::ReadGlobal ||
+             mode == metrics::CollectMode::WriteGlobal)
+      << modeStr;
+  return makeRedirection(*leader, true);
+}
+
+RestStatus RestMetricsHandler::makeRedirection(std::string const& serverId,
+                                               bool last) {
+  auto* pool = server().getFeature<NetworkFeature>().pool();
+  if (pool == nullptr) {
+    THROW_ARANGO_EXCEPTION(TRI_ERROR_SHUTTING_DOWN);
   }
-  return RestStatus::DONE;
+
+  network::RequestOptions options;
+  options.timeout = network::Timeout(30.0);
```

Comment:
Hardcoded timeout :(

Reply:
Its old code, maybe I should change it?

Reply:
I'd prefer to have a configurable timeout. But if it is too much effort we could do that later together with other places.